waiter/src/waiter/process_request.clj

```diff
@@ -483,6 +482,57 @@
             (meters/mark! (metrics/waiter-meter "core" "process-errors"))
             (utils/exception->response e request)))))))
 
+(defn service-id->interstitial-promise
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id]
+  (get @service-id->interstitial-promise-atom service-id))
+
+(defn start-service-interstitial!
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id interstitial-secs]
+  (if-let [existing-interstitial-timed-out-promise
+           (service-id->interstitial-promise service-id->interstitial-promise-atom service-id)]
+    existing-interstitial-timed-out-promise
+    (let [interstitial-timed-out-promise (promise)]
+      (->> (fn [service-id->interstitial-promise]
+             (log/info "creating interstitial promise for" service-id)
+             (if-not (contains? service-id->interstitial-promise service-id)
+               ;; TODO (shams) trigger call to start service in follow-up PR, so we can move it out of prepare instance
+               (assoc service-id->interstitial-promise service-id interstitial-timed-out-promise)
+               service-id->interstitial-promise))
+           (swap! service-id->interstitial-promise-atom))
+      ;; TODO (shams) trigger deliver as soon as we have a healthy instance for this service,
+      ;; TODO (shams) e.g. by running a separate go-block that listens to scheduler messages
+      (if (pos? interstitial-secs)
+        (async/go
+          (-> interstitial-secs t/seconds t/in-millis async/timeout async/<!)
+          (log/info "interstitial timed out for" service-id)
+          (deliver interstitial-timed-out-promise :timed-out))
+        (deliver interstitial-timed-out-promise :opted-out))
+      interstitial-timed-out-promise)))
+
+(defn remove-service-interstitial!
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id]
+  (log/info "removing interstitial promise for" service-id)
+  (swap! service-id->interstitial-promise-atom dissoc service-id))
+
+(defn wrap-interstitial
+  "TODO (shams) docstring"
+  [handler service-id->interstitial-promise-atom]
+  (fn wrap-interstitial-handler [{:keys [descriptor] :as request}]
+    (let [{:keys [service-description service-id]} descriptor
+          ;; trigger interstitial-promise irrespective of the bypass parameter
+          interstitial-promise (or (service-id->interstitial-promise service-id->interstitial-promise-atom service-id)
+                                   (let [{:strs [interstitial-secs]} service-description]
+                                     (start-service-interstitial! service-id->interstitial-promise-atom service-id interstitial-secs)))]
+      (if (or (comment TODO contains-bypass-interstitial-param)
+              (realized? interstitial-promise))
+        (handler request)
```

Comment:
`start-new-app` is called in `handler`.  Are we planning on have interstitial start the app itself?  Are we going to have two places where apps might be started?

Reply:
The plan is to have interstitial be the only one that starts the service.